### ISSUE TITLE
Instance declarations risk loopiness; easy to fix

### ISSUE THREAD

**Issue by simonpj** [diku-dk/futhark#1832]
With UndecidableInstances (which futhark uses a lot) it is all too easy for GHC to make an instance declarations whose superclasses are bottom (they diverge).  This has been the source of many bug reports, and a series of fixes.  I am just completing the latest fix in [GHC MR !7050](https://gitlab.haskell.org/ghc/ghc/-/merge_requests/7050), which closes a loophole in the previous story, discovered in [GHC ticket #20666](https://gitlab.haskell.org/ghc/ghc/-/issues/20666)

Now, it turns out that futhark was (inadvertently) exploing this loophole.  So when it's fixed, you'll need to update futhark.  A patch is below; it is backward compatible and will work with older GHCs).

If you are interested in the details, the above links will help; plus the new user manual section "Undecidable instances and loopy superclasses" in the user manual. Or you might want to read `Note [Recursive superclasses]` and `Note [Solving superclass constraints]` in GHC's source code [GHC.Tc.TyCl.Instance](https://gitlab.haskell.org/ghc/ghc/-/blob/master/compiler/GHC/Tc/TyCl/Instance.hs).

The [patch is here](https://gitlab.haskell.org/ghc/ghc/-/merge_requests/7050#note_470835).

**Comment by athas** [comment]
Interesting!  And not terribly surprising.  I do recall some seemingly simple type errors occasionally sending GHC into an infinite loop.  This might be the cause.  Lovely that GHC will help steer the gun slightly further from our feet in the future.

**Comment by athas** [comment]
Unfortunately the suggested patch seems to make GHC 9.2.4 (which we use in CI) go into an infinite loop when compiling Futhark.IR.GPU.  We have a little while before the next GHC comes out, so maybe we should take the time to shrink our usage of UndecidableInstances.

**Comment by simonpj** [comment]
> Unfortunately the suggested patch seems to make GHC 9.2.4 (which we use in CI) go into an infinite loop when compiling Futhark.IR.GPU

Gah!  That's a bug in 9.2.4 -- GHC should never loop.  I will try to repro but meanwhile if anyone feels able to distil a repro case and file a bug against 9.2.4 that would be super helpful

**Comment by athas** [comment]
It compiles fine with GHC 9.4.3 (although with a few incorrect constraint warnings in the patched instances), so I think we will just postpone this patch until downstream has moved to that version.

**Comment by simonpj** [comment]
OK, that's a relief.  I'm disinclined to debug 9.2.   

Yes, the warnings will be there in 9.4 but will go away once this patch lands.    All good!

**Comment by athas** [comment]
Futhark now compiles fine with 9.6 (once you convince cabal to relax various bounds in dependencies), so closing this.
